alert-common/src/main/java/com/synopsys/integration/alert/common/workflow/processor2/MessageContentCollapser.java

```diff
@@ -0,0 +1,128 @@
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.synopsys.integration.alert.common.workflow.processor2;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.EnumMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import com.synopsys.integration.alert.common.enumeration.ItemOperation;
+import com.synopsys.integration.alert.common.exception.AlertException;
+import com.synopsys.integration.alert.common.message.model.ComponentItem;
+import com.synopsys.integration.alert.common.message.model.ComponentKey;
+import com.synopsys.integration.alert.common.message.model.LinkableItem;
+import com.synopsys.integration.alert.common.message.model.ProviderMessageContent;
+
+@Component
+public class MessageContentCollapser {
+    private final Map<ItemOperation, BiFunction<Map<ComponentKey, ComponentItem>, ComponentItem, Void>> operationFunctionMap;
+
+    @Autowired
+    public MessageContentCollapser() {
+        final BiFunction<Map<ComponentKey, ComponentItem>, ComponentItem, Void> addFunction = createAddFunction();
+        final BiFunction<Map<ComponentKey, ComponentItem>, ComponentItem, Void> deleteFunction = createDeleteFunction();
+        operationFunctionMap = new EnumMap<>(ItemOperation.class);
+        operationFunctionMap.put(ItemOperation.ADD, addFunction);
+        operationFunctionMap.put(ItemOperation.UPDATE, addFunction);
+        operationFunctionMap.put(ItemOperation.DELETE, deleteFunction);
+    }
+
+    public List<ProviderMessageContent> collapse(final List<ProviderMessageContent> messages) {
+        final List<ProviderMessageContent> collapsedMessages = new ArrayList<>(messages.size());
+        for (final ProviderMessageContent message : messages) {
+            final Map<ComponentKey, ComponentItem> categoryDataCache = new LinkedHashMap<>();
+            message.getComponentItems().forEach(item -> processOperation(categoryDataCache, item));
+
+            final Optional<ProviderMessageContent> collapsedContent = rebuildTopic(message, categoryDataCache.values());
+            if (collapsedContent.isPresent()) {
+                collapsedMessages.add(collapsedContent.get());
+            }
+        }
+
+        return collapsedMessages;
+    }
+
+    private BiFunction<Map<ComponentKey, ComponentItem>, ComponentItem, Void> createAddFunction() {
+        return (categoryDataCache, categoryItem) -> {
+            categoryDataCache.put(categoryItem.getComponentKey(), categoryItem);
+            return null;
```

Comment:
Does the BiFunction require a return?

Reply:
the return type is void so returning null fulfills that part of the contract.

Reply:
Exactly. Also, this is mostly just copied/pasted from the old logic.